### ISSUE TITLE
Nice to have settings: use semicolon, space before unary tag ending

### ISSUE THREAD

**Issue by gls7** [jcberquist/commandbox-cfformat#103]
We had a rogue developer stop using semicolons just because he could. Would be nice to get those fixed up. Maybe have a use semicolons setting?

Also, our preference is <cfset blah = 1 /> vs <cfset blah = 1/>. I noticed that the <cfreturn /> didn't become <cfreturn/>.

And one more: Spaces after a comma in lists of params. Not string lists.

Would be nice to see settings for those if it's not too much work.

**Comment by jcberquist** [comment]
With regard to the first request, insertion of semicolons is not currently possible. The reason is that cfformat does not perform a full parse into an AST or even just enough to know where statements end, so it does not "know" about full statements that should or could be terminated with a semicolon. It is possible down the road that this could be improved.

I don't think I follow your other requests. Could you clarify them with examples?

**Comment by gls7** [comment]
I just noticed the second line in my initial post became something different, it swallowed up my markup.

Our preference is `<cfset blah = 1 />` vs `<cfset blah = 1/>`. I noticed that the `<cfreturn />` didn't become `<cfreturn/>.`
It seems that your default behavior is to remove any space before the end of a unary tag like `<cfset blah = 1/>` but we have as our convention to put a space before the unary tag ending `<cfset blah = 1 />`. When I run our code through the formatter we lose that extra space at the end. But I didn't see it being lost on unary tags with no other attributes/values. For example `<cfreturn />` didn't lose the space.

We prefer spaces after a comma in lists of params just as in written english.
`someFunction(a, b, c)`
is preferred over
`someFunction(a,b,c)`
Basically any comma-separated list of parameters should have spacing after the comma.

But of course string lists should be left untouched. `blah = "1,2,3"` and `blah = "a, b, c"` should be left alone.

**Comment by jcberquist** [comment]
Ok, great. The tag examples are helpful. I believe I can improve that. With regard to the function call arguments I believe that spaces after commas are always currently inserted - are you seeing something different? And string lists are definitely left alone, the contents of strings are not touched (except to handle switching between single and double quotes).

**Comment by gls7** [comment]
I'll find examples that you can test with.

**Comment by gls7** [comment]
Here is an example file that has been formatted 1, 2, and 3 times.

[formatting.zip](https://github.com/jcberquist/commandbox-cfformat/files/6031234/formatting.zip)

**Comment by gls7** [comment]
With regard to the function call arguments, I can't see it happening now. If it comes up again later on I'll open a new issue.

**Comment by gls7** [comment]
> Here is an example file that has been formatted 1, 2, and 3 times.
> 
> [formatting.zip](https://github.com/jcberquist/commandbox-cfformat/files/6031234/formatting.zip)

Hey I was curious if you have had time to look at the tabbing issue when the same code is run through the formatter more than once. This is what's holding us back from using this awesome code.

**Comment by jcberquist** [comment]
Have you checked the formatting recently? I see the difference between the files you shared, but I can't reproduce the spacing issue myself. Perhaps it has to do with specific formatting settings that you use?

**Comment by gls7** [comment]
Pulled and tested. I don't see the white space problem now.
